iot/end_to_end_example/device/app/src/main/kotlin/com/example/androidthings/coffeecontrol/CoffeeControlActivity.kt

```diff
@@ -0,0 +1,324 @@
+        Log.d(TAG, if (deviceDisplay != null) "Initialized LED strip" else "Error initializing LED strip")
+        deviceLedstrip?.brightness = LEDSTRIP_BRIGHTNESS
+        for (i in deviceRainbow.indices) {
+            deviceRainbow[i] = Color.rgb(0, 0, 0)
+        }
+
+        deviceHandler.post(deviceAnimateRunnable)
+
+        // GPIO led
+        val pioManager = PeripheralManager.getInstance()
+        deviceLed = pioManager.openGpio(BoardDefaults.ledGpioPin)
+        deviceLed.setEdgeTriggerType(Gpio.EDGE_NONE)
+        deviceLed.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
+        deviceLed.setActiveType(Gpio.ACTIVE_HIGH)
+
+        // Configure the Cloud IoT Connector --
+        val pkId = resources.getIdentifier("privatekey", "raw", packageName)
+        try {
+            if (pkId != 0) {
+                val privateKey = applicationContext
+                        .resources.openRawResource(pkId)
+                val keyBytes = privateKey.readBytes()
+
+                val spec = PKCS8EncodedKeySpec(keyBytes)
+                val kf = KeyFactory.getInstance("RSA")
+                val keys = KeyPair(null, kf.generatePrivate(spec))
+
+                // Configure Cloud IoT Core project information
+                val connectionParams = ConnectionParams.Builder()
+                    .setProjectId(BuildConfig.PROJECT_ID)
+                    .setRegistry(BuildConfig.REGISTRY_ID, BuildConfig.ZONE)
+                    .setDeviceId(BuildConfig.DEVICE_ID)
+                    .build()
+
+                // Initialize the IoT Core client
+                client = IotCoreClient.Builder()
+                    .setConnectionParams(connectionParams)
+                    .setKeyPair(keys)
+                    .setOnConfigurationListener(OnConfigurationListener {
+                        this.onConfigurationReceived(it)
+                    })
+                    .build()
+
+                // Connect to Cloud IoT Core
+                client.connect()
+
+                deviceHandler.post(devieTempReportRunnable)
+            }
+        } catch (ikse: InvalidKeySpecException) {
+            Log.e(TAG, "INVALID Key spec", ikse)
+        } catch (nsae: NoSuchAlgorithmException) {
+            Log.e(TAG, "Algorithm not supported", nsae)
+        } catch (ioe: IOException) {
+            Log.e(TAG, "Could not load key from file", ioe)
+        }
+    }
+
+    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+        if (keyCode == KeyEvent.KEYCODE_A) {
+            Log.d(TAG, "A Click")
+            return true
+        }
+        return super.onKeyUp(keyCode, event)
+    }
+
+    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
+        return if (keyCode == KeyEvent.KEYCODE_A) {
+            true
+        } else super.onKeyUp(keyCode, event)
+    }
+
+    override fun onDestroy() {
+        deviceButtonInputDriver.close()
+
+        // deviceLedstrip is optional
```

Comment:
Can you use `?.let`? 

This can be written as:
```
        deviceLedstrip?.let { device ->
            devic.brightness = 0
            device.write(IntArray(7))
            device.close()
        }
```